src/github.com/matrix-org/go-neb/realms/github/github.go

```diff
@@ -60,13 +77,85 @@ func (r *githubRealm) RequestAuthSession(userID string, req json.RawMessage) int
 	}{u.String()}
 }
 
-func (r *githubRealm) AuthSession(userID, realmID string) types.AuthSession {
+func (r *githubRealm) OnReceiveRedirect(w http.ResponseWriter, req *http.Request) {
+	// parse out params from the request
+	code := req.URL.Query().Get("code")
+	state := req.URL.Query().Get("state")
+	logger := log.WithFields(log.Fields{
+		"state": state,
+	})
+	logger.WithField("code", code).Print("GithubRealm: OnReceiveRedirect")
+	if code == "" || state == "" {
+		failWith(logger, w, 400, "code and state are required", nil)
+		return
+	}
+	// load the session (we keyed off the state param)
+	session, err := database.GetServiceDB().LoadAuthSessionByID(r.ID(), state)
+	if err != nil {
+		// most likely cause
+		failWith(logger, w, 400, "Provided ?state= param is not recognised.", err)
+		return
+	}
+	ghSession := session.(*githubSession)
```

Comment:
It's probably unnecessary, but it might be polite to use the form of type assertion that doesn't panic.